pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1246,6 +1278,52 @@ func (r *ReconcileClusterDeployment) deleteJobOnHashChange(existingJob, generate
 	return newJobNeeded, nil
 }
 
+// cleanupInstallLogPVC will immediately delete the PVC (should it exist) if the cluster was installed successfully, without retries.
+// If there were retries, it will delete the PVC if it has been more than 7 days since the job was completed.
+func (r *ReconcileClusterDeployment) cleanupInstallLogPVC(cd *hivev1.ClusterDeployment, existingJob *batchv1.Job, cdLog log.FieldLogger) error {
+	if !cd.Status.Installed {
+		return nil
+	}
+
+	if existingJob == nil {
+		return nil
+	}
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	// PVC re-uses the name of the install job as they are related, and we know it should be unique.
+	err := r.Get(context.TODO(), types.NamespacedName{Name: existingJob.Name, Namespace: existingJob.Namespace}, pvc)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		cdLog.WithError(err).Error("error looking up install logs PVC")
+		return err
+	}
+
+	if cd.Status.InstallRestarts == 0 {
+		cdLog.WithField("pvc", pvc.Name).Info("deleting logs PersistentVolumeClaim for installed cluster with no restarts")
+		if err := r.Delete(context.TODO(), pvc); err != nil {
+			cdLog.WithError(err).Error("error deleting install logs PVC")
+			return err
+		}
+		return nil
+	}
+
+	// Otherwise, delete if more than 7 days have passed.
+	if time.Since(existingJob.Status.CompletionTime.Time) > (7 * 24 * time.Hour) {
```

Comment:
I would like to see this as a test of the status condition on the clusterdeployment rather than the job completion time. It is possible that the job may have been deleted by the user in the 7 days since it completed. The controller needs to be resilient to this possibility. The other option is to immediately delete the PVC when the job no longer exists. In either case, the controller cannot rely on getting the name of the PVC from the name of the job.

This is another item in the column in favor of separating the creation of the PVC from the creation of the job.

Reply:
We don't seem to have any such condition, I think I will add a Status.InstalledTimestamp, have been thinking about recording some other timestamps such as DNS resolution start/stop times for better metrics.